lib/baw-workers/partial_payload.rb

```diff
@@ -0,0 +1,151 @@
+        existing_base = get(key)
+        if existing_base
+          raise InconsistentBasePayload if existing_base != BawWorkers::ResqueJobId::normalise(base_payload)
+
+          # return a hash with the absolute redis_key that can be used to retrieve the base payload
+          {PARTIAL_PAYLOAD_KEY.to_sym => @communicator.add_namespace(add_namespace(key))}
+        else
+          create(base_payload, key)
+        end
+      end
+
+      # Reconstruct a payload by combining a partial payload with its base payload.
+      # resolve can reconstruct several nested payload partials via recursion. Its only limitations
+      # are callstack size and Redis performance.
+      # Payloads that are not hashes or are hashes without the partial payload key, are ignored and returned unmodified.
+      # @param [Hash] payload
+      # @return [Hash] the reconstructed payload
+      def resolve(payload)
+        return payload unless payload.is_a?(Hash)
+
+        has_string = payload.has_key?(PARTIAL_PAYLOAD_KEY)
+        has_symbol = payload.has_key?(PARTIAL_PAYLOAD_KEY.to_sym)
+        if has_string || has_symbol
+          # we assume the full redis key (with namespaces) has been stored
+          redis_key = payload.delete(has_string ? PARTIAL_PAYLOAD_KEY : PARTIAL_PAYLOAD_KEY.to_sym)
+
+
+          base = @communicator.get(redis_key, {no_namespace: true})
+
+          if base.nil?
+            raise PartialPayloadMissing.new("Could not retrieve partial payload `#{redis_key}`")
+          end
+
+          # RECURSIVE - allow the base payload to have its own base payload
+          new_base = resolve(base)
+
+          payload = new_base.merge(payload)
+        end
+
+        payload
+      end
+
+      # Gets a base payload if it exists
+      # @param [String] key - they key for the base payload to retrieve
+      # @return [Hash] returns a Hash if found, otherwise nil
+      def get(key)
+        key = add_namespace(key)
+        @communicator.get(key)
+      end
+
+      # Delete the base payload.
+      # Does not cascade and delete other linked base payloads.
+      # @param [String] key - they key to delete
+      # @return [Boolean] True if the delete succeeded.
+      def delete(key)
+        key = add_namespace(key)
+
+        @communicator.delete(key)
+      end
+
+      # Delete the base payload.
+      # *Does* cascade and delete other linked base payloads.
+      # Throws `PartialPayloadMissing` if a linked base payload can not be found.
+      # @param [String] key - they key to delete
+      # @return [Boolean] True if the delete succeeded.
+      def delete_recursive(key)
+        has_namespace = key.include?(':' + REDIS_NAMESPACE + ':')
+        key = add_namespace(key) unless has_namespace
+
+        # first get the partial payload
+        partial = @communicator.get(key, {no_namespace: has_namespace})
+
+        if partial.nil?
+          raise PartialPayloadMissing.new("Could not retrieve partial payload `#{key}`")
```

Comment:
Yes... but if it can't be found it is already in an inconsistent state. Any other choices?